mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -32,35 +32,81 @@ pub fn magic_dom(input: TokenStream) -> TokenStream {
     gen.to_string().parse().unwrap()
 }
 
+fn gen_constructor(inherit: bool,
+                   num_fields: u32,
+                   name: &quote::Ident,
+                   js_class: &quote::Ident,
+                   get_callargs: &Vec<quote::Tokens>,
+                   setter_invocations: &Vec<quote::Tokens>) -> quote::Tokens {
+    if inherit {
+        quote!{}
+    } else {
+        let constructor = quote::Ident::from(format!("{}_constructor", name));
+        let from_obj_err = quote::Ident::from(format!("b\"Fail to construct {} from JS \
+                                                       object\\0\" as *const u8 as *const \
+                                                       libc::c_char", name));
+        let arg_num_err = quote::Ident::from(format!("b\"constructor requires exactly {} \
+                                                      arguments\\0\".as_ptr() as *const \
+                                                      libc::c_char", num_fields));
+        quote! {
+            #[allow(non_snake_case)]
+            pub unsafe extern "C" fn #constructor(cx: *mut JSContext, argc: u32, vp: *mut JS::Value) -> bool {
+                let call_args = CreateCallArgsFromVp(argc, vp);
+                if call_args._base.argc_ != #num_fields {
+                    JS_ReportErrorASCII(cx, #arg_num_err);
+                    return false;
+                }
+
+                #(#get_callargs)*
+
+                rooted!(in(cx) let jsobj = jsapi::JS_NewObjectForConstructor(cx,
```

Comment:
I think we need to switch from generating `JSClass`es and start generating `js::Class`es so we can define `js::ClassSpec`s and the `createPrototype` hooks. That will allow us to create the proper prototype chains, so that we get the right inheritance here.